frontend/src/features/user/billing/BillCharges/components/BillingTable.tsx

```diff
@@ -0,0 +1,186 @@
+import { useEffect, useMemo } from 'react'
+import {
+  Column,
+  useFlexLayout,
+  usePagination,
+  useResizeColumns,
+  useSortBy,
+  useTable,
+} from 'react-table'
+import { Flex, Icon, Table, Tbody, Td, Th, Thead, Tr } from '@chakra-ui/react'
+
+import { FormAuthType, FormBillingStatistic } from '~shared/types'
+
+import { BxsChevronDown } from '~assets/icons/BxsChevronDown'
+import { BxsChevronUp } from '~assets/icons/BxsChevronUp'
+
+type BillingColumnData = Pick<
+  FormBillingStatistic,
+  'formName' | 'adminEmail' | 'authType' | 'total'
+>
+
+const AUTHTYPE_TO_TEXT = {
+  [FormAuthType.NIL]: '-',
+  [FormAuthType.SP]: 'Singpass',
+  [FormAuthType.SGID]: 'sgID',
+  [FormAuthType.MyInfo]: 'MyInfo',
+  [FormAuthType.CP]: 'Corppass',
+}
+
+const BILLING_TABLE_COLUMNS: Column<BillingColumnData>[] = [
+  {
+    Header: 'Form title',
+    accessor: 'formName',
+    sortType: 'basic',
+    minWidth: 300, // minWidth is only used as a limit for resizing
+    width: 400, // width is used for both the flex-basis and flex-grow
+    maxWidth: 500, // maxWidth is only used as a limit for resizing
+  },
+  {
+    Header: 'Owner',
+    accessor: 'adminEmail',
+    sortType: 'basic',
+    minWidth: 150,
+    width: 250,
+    maxWidth: 350,
+  },
+  {
+    Header: 'Authentication',
+    accessor: (row) => AUTHTYPE_TO_TEXT[row.authType],
+    sortType: 'basic',
+    minWidth: 150,
+    width: 160,
+    maxWidth: 170,
+  },
+  {
+    Header: '# of logins',
+    accessor: 'total',
+    sortType: 'number',
+    minWidth: 130,
+    width: 140,
+    maxWidth: 150,
+  },
+]
+
+export const BillingTable = ({
+  loginStats,
+  currentPage,
+}: {
+  loginStats: FormBillingStatistic[]
+  currentPage: number
+}) => {
+  const {
+    prepareRow,
+    getTableProps,
+    getTableBodyProps,
+    headerGroups,
+    page,
+    gotoPage,
+  } = useTable<BillingColumnData>(
+    {
+      columns: useMemo(() => BILLING_TABLE_COLUMNS, []),
+      data: useMemo(() => loginStats, [loginStats]),
```

Comment:
These 2 `useMemo()` are not needed
```suggestion
      columns: BILLING_TABLE_COLUMNS,
      data: loginStats,
```

Reply:
that's what I thought too, but it turns out if you get rid of these `useMemo` calls, it causes an infinite render loop for some reason :(

see also: https://react-table-v7.tanstack.com/docs/quick-start#:~:text=It%27s%20important%20that%20we%27re%20using%20React.useMemo%20here%20to%20ensure%20that%20our%20data%20isn%27t%20recreated%20on%20every%20render.%20If%20we%20didn%27t%20use%20React.useMemo%2C%20the%20table%20would%20think%20it%20was%20receiving%20new%20data%20on%20every%20render%20and%20attempt%20to%20recalculate%20a%20lot%20of%20logic%20every%20single%20time.%20Not%20cool!

Reply:
In that reference link you pasted, the data and column definition are objects and arrays created in-line, so it's a specific case. Objects/Arrays created in-line will always have different references (basically theres one new structure created per function call), and that's what causes them to be seen as different even though the structure doesn't change. 

In the code you have here, `BILLING_TABLE_COLUMNS` is a module constant, so not only the structure is static, but the reference is static too, and there's no need for an additional `useMemo()` wrapper. 

The same applies to `loginStats`.

Basically a `useMemo()` like this:
```
useMemo(() => someRef, [someRef])
```
is useless by definition 😅 . `useMemo()` returns the same data reference across multiple calls as long as the list of depencies hasn't changed. Obviously if `someRef` changes, `useMemo()` will re-momoize, which in this case will store and return `someRef` again, so you can just use `someRef` directly.

Reply:
Actually, you're right - I just tested and it works! 

Previously (before I added the `useMemo`s) I was getting an infinite render loop and somehow adding those `useMemo`s fixed it... But now, it still works when I remove the `useMemo`s? Not too sure why that happened haha

Upon further reflection upon what you said above - said infinite render loop occured in the storybook, not on the app itself. Quite sure that back then, the mocked data I input was inline into the storybook args, but since then I've moved the mock data out of that file and into an msw file. That must be why it used to throw that loop. tagging @karrui for reference haha

Regardless, will make this change! Thanks for the explanation Tim 😅